### ISSUE TITLE
Complete English translation

### ISSUE THREAD

**Issue by schaafjs** [Cay-Zhang/RSSBud#22]
([Chinese translation below](#chinese))

Hi there,
thank you very much for developing and providing this app. Would you be able to fully translate it to english? I have attached two screenshots, one showing missing translations for the different feed types and one showing missing translations in the "Shortcut Workshop" section in the settings.

## Missing translations for different feed types
<img src="https://user-images.githubusercontent.com/44237133/161427106-52652635-816d-488a-a600-e634998ebe21.PNG" height="700">

## Missing translations in "Shortcut Workshop"
<img src="https://user-images.githubusercontent.com/44237133/161427113-79f50176-1dde-41af-a4f1-69ae52627a67.PNG" height="700">

If I can be of any assistance during translation or for proofreading, feel free to ask!

Thank you very much.

# Chinese
<a name="chinese"></a>

你们好。
非常感谢你开发和提供这个应用程序。你能不能把它完全翻译成英文？我附上了两张截图，一张显示不同饲料类型的翻译缺失，一张显示设置中的 "快捷工作室 "部分的翻译缺失。

## 缺少不同类型的翻译
<img src="https://user-images.githubusercontent.com/44237133/161427106-52652635-816d-488a-a600-e634998ebe21.PNG" height="700">

## "捷径工作室 "中缺少的翻译
<img src="https://user-images.githubusercontent.com/44237133/161427113-79f50176-1dde-41af-a4f1-69ae52627a67.PNG" height="700">

如果我在翻译过程中或在校对方面能提供任何帮助，请随时提出!

非常感谢你。

**Comment by Cay-Zhang** [comment]
Hey,

English translation for the shortcut workshop should come with v2 which is still under development (very slow though).

For RSSHub feeds, I don't really have any control over translations since RSSBud relies on the rule file provided by the RSSHub project (learn more about it [here](https://docs.rsshub.app/en/joinus/quick-start.html#submit-new-rsshub-radar-rule)). I did propose some solutions for translating RSSHub rule titles in their Telegram group but sadly nothing came out of it.

By the way, in v2 the path for an RSSHub feed will be shown in conjunction with its title so hopefully, non-Chinese speakers could understand its meaning a little better.

**Comment by schaafjs** [comment]
Hi,

thanks for your reply. Couldn't I also just localize the file you linked? Since it's JSON, one could just automatically translate the title property, yielding proper English localization of the displayed rules. This would mean that you'd have to be able to set a custom URL for the rules, but apart from that it should work, right?
Localization should be possible via GitHub Actions plus some translation API. The [one offered by DeepL](https://www.deepl.com/en/pro-api) has a quite generous free tier of 500k characters a month, for example.

If you'd consider making the rules URL a custom setting, I can have a look at the translation part.

While this doesn't really solve the whole issue at hand and wouldn't directly contribute to the upstream project, I still think it is the best bet for now.

**Comment by Cay-Zhang** [comment]
Yes, I am planning on supporting custom rule sources, with an expanded schema and support for multiple rules!

Having a different rule file automatically generated with translated titles sounds pretty cool since it doesn't require a schema change from the RSSHub side. It would be much appreciated if you could help create a GitHub action that translates [rule](https://rsshub.js.org/build/radar-rules.js) titles into an arbitrary language.

**Comment by schaafjs** [comment]
> It would be much appreciated if you could help create a GitHub action that translates [rule](https://rsshub.js.org/build/radar-rules.js) titles into an arbitrary language.

I'll do some research into what's the best way to go about this and then look into implementing it.

I think a separate repository to house the initial rules file and the derived translations plus the final GitHub action would probably be the best place, right?

**Comment by Cay-Zhang** [comment]
Thanks for your help! I've opened up [a new repo for RSSBud rules](https://github.com/Cay-Zhang/RSSBudRules). Feel free to contribute there. I'll be designing the extended schema in the meantime.

**Comment by schaafjs** [comment]
Sorry for the late reply. I was quite busy in the last week and will take a look at the translations in the coming days.

It would probably make sense to update the ruleset in that repo just like in this one using the [update-rules workflow](https://github.com/Cay-Zhang/RSSBud/blob/main/.github/workflows/update-rules.yml).

**Comment by Cay-Zhang** [comment]
Thanks for reminding me! Just added the workflow.

**Comment by schaafjs** [comment]
Hi there,

I can provide a small update to progress on the autotranslation. I have checked out a couple of different translation tools but have not found one that is fully sufficient as of yet.
Two somewhat final candidates for now are the following:
- [KhaledMohamedP/translate-json-object](https://github.com/KhaledMohamedP/translate-json-object): seems to work well, with three major caveats:
  1. An API key for the supported services (Google and Yandex) is needed. This means that we might run into usage limitations, especially during the initial translation.
  2. Translations are not incremental, without additional logic every word will be retranslated every time. Git will easily handle this on the version control side of things but this will most likely exhaust every free quota available to as quite fast.
  3. Translations can not be limited to specific keys of the `json` file. The value of each key will be translated. This does not work welll with almost all URLs.
- [machine-translator](https://github.com/marketplace/actions/machine-translator) by @IEvangelist: Issues i. and ii. apply here as well, Azure Cognitive Services are used. They offer a free tier.

  Only translating predefined keys [has been requested](https://github.com/IEvangelist/resource-translator/issues/35), but it does not seem that it is actively worked on.

For now we have to either wait or program our own solution it seems like.

Considering the later option, we might be lucky using [LibreTranslate](https://github.com/LibreTranslate/LibreTranslate) which has [Mirrors](https://github.com/LibreTranslate/LibreTranslate#mirrors) that do not require an API key.
Would you mind checking out their translation quality? Translations can be made using their [web interface](https://libretranslate.com/).

Btw, it would probably be more fitting to continue this conversation over in the new repo, you can [transfer the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository) if you wish.
